pkg/events/derive/symbols_collision.go

```diff
@@ -0,0 +1,288 @@
+// It updates the given SO if symbols are missing.
+func (soColGen *SymbolsCollisionArgsGenerator) findSOCollisions(so *loadingSOInstance, loadedSO sharedobjs.ObjInfo) ([]string, error) {
+	var err error
+	collisions, ok := soColGen.collisionsCache.GetCollision(so.Id, loadedSO.Id)
+	if !ok {
+		if so.ExportedSymbols == nil {
+			so.ExportedSymbols, err = soColGen.soLoader.GetExportedSymbols(so.ObjInfo)
+			so.FilterSymbols(soColGen.symbolsBlackList)
+			so.FilterOutSymbols(soColGen.symbolsWhiteList)
+			if err != nil {
+				return nil, err
+			}
+		}
+		lsoSyms, err := soColGen.soLoader.GetExportedSymbols(loadedSO)
+		if err != nil {
+			return nil, err
+		}
+		lso := loadingSOInstance{ObjInfo: loadedSO, ExportedSymbols: lsoSyms}
+		lso.FilterSymbols(soColGen.symbolsBlackList)
+		lso.FilterOutSymbols(soColGen.symbolsWhiteList)
+		collisions = so.GetCollisions(lso)
+		soColGen.collisionsCache.AddCollisions(so.Id, loadedSO.Id, collisions)
+	}
+	return collisions, nil
+}
+
+// processesLoadedObjectsCache is a cache for shared objects loaded to a process
+type processesLoadedObjectsCache struct {
+	cache *simplelru.LRU
+}
+
+// GetProcessLoadedObjects get the shared SOs loaded to the process, and if the process existed in the cache.
+func (pcache *processesLoadedObjectsCache) GetProcessLoadedObjects(pid int) ([]sharedobjs.ObjInfo, bool) {
+	var loadedObjects []sharedobjs.ObjInfo
+	loadedObjectsIface, ok := pcache.cache.Get(pid)
+	if ok {
+		loadedObjects = loadedObjectsIface.([]sharedobjs.ObjInfo)
+		return loadedObjects, true
+	} else {
+		return loadedObjects, false
+	}
+}
+
+func (pcache *processesLoadedObjectsCache) SetProcessLoadedObjects(pid int, loadedObjects []sharedobjs.ObjInfo) {
+	pcache.cache.Add(pid, loadedObjects)
+}
+
+// sharedObjectsCollisionsCache is a cache for collision checks performed already
+// The results are saved per SO, which means that each collision is saved twice (because in each collision there are
+// 2 SOs which collide)
+type sharedObjectsCollisionsCache struct {
+	cache *simplelru.LRU
+}
+
+// AddCollisions add the collisions between 2 shared objects to the cache.
+func (socCache sharedObjectsCollisionsCache) AddCollisions(obj1 sharedobjs.ObjID,
+	obj2 sharedobjs.ObjID,
+	collisions []string) {
+	socCache.addObjCollisions(obj1, obj2, collisions)
+	socCache.addObjCollisions(obj2, obj1, collisions)
+}
+
+// GetCollision return the symbols collided between 2 shared objects from the cache if exist.
+// If the the collisions are cached for one object, will update it in the seconde object cache.
+// Return bool indication if the collision existed in the cache.
+func (socCache sharedObjectsCollisionsCache) GetCollision(obj1 sharedobjs.ObjID,
+	obj2 sharedobjs.ObjID) ([]string, bool) {
+	obj1Collisions, ok := socCache.getObjCollisions(obj1)
+	if ok {
+		collisions, ok := obj1Collisions[obj2]
+		if ok {
+			socCache.addObjCollisions(obj2, obj1, collisions)
```

Comment:
So your cache is always duplicated (also in mem footprint) and you're workarounding the fact that one of the duplicates might have vanish because of the LRU. Couldn't you have made a:

map[struct{obj1,obj2}][]string

and do lookups with {obj1,obj2} and {obj2,obj1} (or even sort obj1, obj2 for a single lookup) ?

Reply:
You are right, this is a lot smarter :)